### ISSUE TITLE
mqtt "Timed out as no activity"

### ISSUE THREAD

**Issue by kglowacki** [aws/aws-iot-device-sdk-java#17]
For a couple of weeks now I've been fighting with AWS IoT to get it to work somewhat reliably - w/o success so far. My sensor is publishing results every 15 minutes or so via MQTT over TLS. After a few hours I get:

10:48:04.467 [MQTT Ping: pi] ERROR o.e.p.c.m.i.ClientState - pi: Timed out as no activity, keepAlive=30,000 lastOutboundActivity=1,486,032,454,464 lastInboundActivity=1,486,032,424,570 time=1,486,032,484,445 lastPing=1,486,032,454,464
10:48:04.516 [pool-1-thread-1] INFO  c.a.s.i.c.c.AwsIotConnection - Connection temporarily lost
10:48:04.521 [pool-1-thread-1] INFO  c.a.s.i.c.c.AwsIotConnection - Connection retry cancelled or exceeded maximum retries
10:48:04.525 [pool-1-thread-1] INFO  c.a.s.i.c.c.AbstractAwsIotClient - Client connection closed: pi

After this error, publish() always fails with:

11:01:04.847 [PmMeter] WARN  c.o.s.a.AWSIotDataCollector - cannot publish result
java.util.concurrent.RejectedExecutionException: Task java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask@1c4089b rejected from java.util.concurrent.ScheduledThreadPoolExecutor@1fb9ffb[Terminated, pool size = 0, active threads = 0, queued tasks = 0, completed tasks = 141]
        at java.util.concurrent.ThreadPoolExecutor$AbortPolicy.rejectedExecution(ThreadPoolExecutor.java:2047)

This happens despite awsIotClient.setMaxConnectionRetries(Integer.MAX_VALUE);
Now, I have no idea how to recover from such error except creating a new AwsIotClient instance - because previous executor thread pool was shutdown and it is useless.

Please, test AWS IoT java sdk with long-lasting connections with rare inactivity.
This is a very often scenario for IoT.

**Comment by fengsongAWS** [comment]
Hi @kglowacki ,

Could you try to increase the keepAlive time and see if this still fails? Currently from what I see is 30 seconds. Another thing we can check is to enable the CloudWatch log to see whether the close is initiated by server or client? 

Thanks for your interest in AWS IoT Java SDK.

**Comment by kglowacki** [comment]
Yep, I set keepAlive=60sec two days ago and so far it seems to improve stability.

Although I'm worried that, if I'm understanding this code correctly, in case of such error AwsIotClient  instance is done (connect() will not work because thread pool has been shut down). 

The only way to recover, without rebooting a hardware, is to recreate AwsIotClient and reattach devices to a new instance. 
IMO AwsIotClient should always try to reconnect (<=MaxConnectionRetries), not just die.

**Comment by fengsongAWS** [comment]
I would think so because the error itself indicates it is a permanent error and no more retries. We will keep investigating and improving our SDK and please let us know if you have more questions and comments. Thanks again for using IoT Java SDK.

**Comment by kglowacki** [comment]
Ok, thanks for investigating the problem. 

I'm new to IoT but my feeling is that AwsIotClient should, once created, try to reconnect forever (whatever error happens - no network, timeouts, etc), and maybe just expose some more detailed state information. Sensors and other devices that are not operated by human have no other choice just try to connect to the network until the world or power supply ends.

**Comment by AngusMorton** [comment]
It is critical we solve this bug as it puts the AwsIotService in an unrecoverable state where you can't even detect that you're in an unrecoverable state.

I believe this error should be resolved transparently the SDK, ie. as consumers of the AWS IoT API we shouldn't need to manually initiate the reconnection to the underlying Paho MQTT client.

**Comment by SanyasiraoM** [comment]
@kglowacki @AngusMorton @fengsongAWS Any progress or work arounds on this?

**Comment by priyamalM** [comment]
i had the same issue, our device agent needed an stable aws connection, and we wanted to continue
even without the aws connection.but this error actually freezes the application,
by default `AWSIotMqttClient.connect();` method will try to reconnect for 5 times and after trying for 5 times it will wait for some time and then this error happens.this is how i was able to fix the issue.

you need to create a `custom Mqttclient by extending AWSIotMqttClient` this will allow
to set the` maxConnectionRetries to 0` , with using `AWSIotMqttClient `we are not able to set the 
`maxConnectionRetries to 0`, default value is 5.

```
public class TestMqttClient extends AWSIotMqttClient{

    private final Logger logger = LoggerFactory.getLogger(TestMqttClient.class);

    public TestMqttClient(String clientEndpoint, String clientId, KeyStore keyStore, String keyPassword) {
        super(clientEndpoint, clientId, keyStore, keyPassword);
        this.maxConnectionRetries = 0;
    }

    public TestMqttClient(String clientEndpoint, String clientId, String awsAccessKeyId, String awsSecretAccessKey) {
        super(clientEndpoint, clientId, awsAccessKeyId, awsSecretAccessKey);
        this.maxConnectionRetries = 0;
    }

    public TestMqttClient(String clientEndpoint, String clientId, String awsAccessKeyId, String awsSecretAccessKey, String sessionToken) {
        super(clientEndpoint, clientId, awsAccessKeyId, awsSecretAccessKey, sessionToken);
        this.maxConnectionRetries = 0;
    }


    @Override
    public void onConnectionClosed() {
        logger.warn("mqtt client connection closed");
    }


    @Override
    public void onConnectionSuccess() {
        super.onConnectionSuccess();
        logger.info("client connection successful");
    }

    @Override
    public void onConnectionFailure() {
        logger.error("client connection failed");
    }

}


/**maintaining the connection**/

if(awsIotClient.getConnection().getConnectionStatus().equals(AWSIotConnectionStatus.RECONNECTING)){
   logger.warn("aws mqtt client is trying to reconnect");
  }

if (awsIotClient.getConnection().getConnectionStatus().equals(AWSIotConnectionStatus.CONNECTED)){
  //code here..	
}

if (awsIotClient.getConnection().getConnectionStatus().equals(AWSIotConnectionStatus.DISCONNECTED)) {
   logger.warn("aws mqtt client is disconnected and trying to re establish..");
   awsIotClient.connect();
}
```

**Comment by forbode** [comment]
This  configuration seems to work reliably for me. I am using aws-iot-device-sdk 1.1.1

**Truths**
My service attempts to send a message 1 per minute
My service should never miss the internet connection longer than 3 days

**My Configuration**
  maxConnectionRetries = 20160; // 336 hours or two weeks
  maxRetryDelay = 60000; // retry each minute
  maxOfflineQueueSize = 303840; // max possible messages in 1 week
  numOfClientThreads = 2;
  keepAliveInterval = 300000; // Once every 5 minutes

**Tests**
I've tested this configuration where there is no internet on startup and it queues publish messages then sends them when internet connections again. 

I've tested this when the internet is connected on startup and then abruptly drops. 

It seems that the issue is knowing what the keep alive interval is or should be. In my case I know that I should see a message at least once per minute.
